### PR TITLE
NMS-13116: fix rounding errors in flow sampling

### PR DESCRIPTION
**Pull request by swachter**
This PR fixes rounding errors that might happen when flows are sampled into their windows.

* moves the check for a negative flow duration into the windows assignment function
* changes the sampling logic that calculates the number of bytes that flows contribute to windows
* adds a property based test that checks that the sum of sampled contributions is equal to the number of bytes of a flow

### CODE REVIEW

File: main/src/main/java/org/opennms/nephron/FlowWindows.java [hunk 1]
```diff
@@ -82,9 +87,17 @@ private FlowWindows(final Duration size) {
...
+        if (lastSwitchedMillis < deltaSwitchedMillis) {
+            LOG.warn("Ignoring flow with negative duration - Flow: {}", flow);
+            return Collections.emptyList();
+        }
+
         // Calculate the first and last window since EPOCH (both inclusive) the flow falls into
-        long firstWindow = flow.getDeltaSwitched().getValue() / windowSize;
-        long lastWindow = flow.getLastSwitched().getValue() / windowSize;
+        long firstWindow = deltaSwitchedMillis / windowSize;
+        long lastWindow = lastSwitchedMillis / windowSize;
```
Comment (Naicisum) [inline main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: This condition is handled by Minion now. Dropping flows are not an option.

**Review by Naicisum — CHANGES_REQUESTED** [review]

File: main/src/main/java/org/opennms/nephron/FlowWindows.java (hunk 1 above)
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: Flows are already dropped in the current implementation (cf. https://github.com/OpenNMS/nephron/blob/master/main/src/main/java/org/opennms/nephron/Groupings.java#L144) when corresponding aggregates are calculates.

The change introduced here is that the flows would already be dropped when they are distributed to their windows.

Comment (Naicisum) [reply main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: Then that section needs to be removed as well. Any unnecessary cpu time needs to be eliminated.

Comment (swachter) [reply main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: I agree, that unnecessary work should be avoided, in general. However, we also have to consider robustness. I am not sure if Minions are the only sources of flows.

@fooker Are Minions the only sources of flows? Can the guard `lastSwitched < deltaSwitched` be dropped or maybe moved into OpenNMS?

Comment (fooker) [reply main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: For now, we can assume all flows are processed/created by OpenNMS. If we need save-guards, we can add them there (either, if protocol specific on the parsers running on minions or in general on the adapters running on sentinel).

File: main/src/main/java/org/opennms/nephron/Groupings.java [hunk 2]
```diff
@@ -138,73 +137,62 @@ private Builder(final T value) {
...
+        }
 
-            double effectiveMultiplier = multiplier;
+        private Aggregate aggregatize(final FlowWindows.FlowWindow window, final FlowDocument flow, final String hostname) {
+            double multiplier = 1;
             if (flow.hasSamplingInterval()) {
                 double samplingInterval = flow.getSamplingInterval().getValue();
                 if (samplingInterval > 0) {
-                    effectiveMultiplier *= samplingInterval;
+                    multiplier = samplingInterval;
```
Comment (fooker) [inline main/src/main/java/org/opennms/nephron/Groupings.java:171]: Can we do the upsampling multiplication outside of the `bytesInWindow` function and just pas in (the already multiplied bytes)? This makes it more clear, that this multiplication has nothing to do with the windowing but is soley protocol related (tbh, I don't even know why we transport this, not doing it on sentinel).

Comment (fooker) [reply main/src/main/java/org/opennms/nephron/Groupings.java:171]: I also don't even know why the sampling interval is transported as a double... There is no such thing as "oversampling".
This is not a fault of your change - but a thing we maybe have to revisit in general.

**Review by fooker — CHANGES_REQUESTED** [review]
Just minor nits.
I like the property based testing.

File: main/src/main/java/org/opennms/nephron/FlowWindows.java (hunk 1 above)
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/FlowWindows.java:100]: Dropped the `lastSwitched < deltaSwitched` safe-guard.

File: main/src/main/java/org/opennms/nephron/Groupings.java (hunk 2 above)
Comment (swachter) [reply main/src/main/java/org/opennms/nephron/Groupings.java:169]: done

**Review by fooker — APPROVED** [review]

**Pull request merged**